File: leetcode/medium/Q39_CombinationSum.py
from typing import List

class Solution:
    answer = []
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.answer = []
        self.find(candidates, [], target, 0)
        return self.answer

    def find(self, candidates: List[int], comb: List[int], target: int, idx: int):
        if sum(comb) > target:
            return
        if sum(comb) == target:
            self.answer.append(comb)
            return
        for i in range(idx, len(candidates)):
            comb = comb.copy()
            comb.append(candidates[i])
            self.find(candidates, comb, target, i)

File: src/medium/Q39_CombinationSum.java
package medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Q39_CombinationSum {
    static class Solution {
        List<List<Integer>> answer;
        public List<List<Integer>> combinationSum(int[] candidates, int target) {
            answer = new ArrayList<>();
            find(0, candidates, target, new ArrayList<>());
            return answer;
        }
        public void find(int idx, int[] can, int target, List<Integer> comb){
            int sum = 0;
            for(int c: comb){
                sum += c;
            }
            while(idx < can.length) {
                if(sum + can[idx] == target){
                    List<Integer> temp = new ArrayList<>(comb);
                    temp.add(can[idx]);
                    answer.add(temp);
                }else if(sum + can[idx] < target){
                    List<Integer> temp = new ArrayList<>(comb);
                    temp.add(can[idx]);
                    find(idx, can, target, temp);
                }
                idx += 1;
            }
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        int[] arr = {2,3,5};
        s.combinationSum(arr, 8);
    }
}
